Fix: Bug: Eclair - Error in GUI when open channel #804

kelvinator07 opened the pull request:
This PR Closes #804

Some [ChannelData](https://github.com/jamaljsr/polar/blob/master/src/lib/lightning/eclair/types.ts#L46) properties for the Eclair node was updated for v0.9.0 when calling its [API](https://acinq.github.io/eclair/#channels-2) and this caused a bug when running that version. 

This PR updates the ChannelData properties to be compatible with v0.9.0, v0.8.0 and v0.7.0.

jamaljsr commented:
Thanks for tackling this issue. I have a bit of feedback after testing this locally.

1. This is unrelated to the code and more about housekeeping. It would be great if you could write more in the PR description explaining the rationale behind the changes you've made. What was the root of the problem? Did something change in the daemons or dependencies that caused this issue. This would just be helpful for reviewers, as well as our future selves to understand why specific code was changed.
2. It looks like this works with the latest version of Eclair, but adding older versions to the network is now broken with similar errors. I guess they changed the structure of the API responses with the latest version. It would be great if our code could support both the old and new structures. If it's too much of a pain, then the only other option would be to drop support for the older versions. I prefer to keep the last 2 major versions around for testing, so that would be a last resort.

kelvinator07 commented:
> Thanks for tackling this issue. I have a bit of feedback after testing this locally.
> 
> 1. This is unrelated to the code and more about housekeeping. It would be great if you could write more in the PR description explaining the rationale behind the changes you've made. What was the root of the problem? Did something change in the daemons or dependencies that caused this issue. This would just be helpful for reviewers, as well as our future selves to understand why specific code was changed.
> 2. It looks like this works with the latest version of Eclair, but adding older versions to the network is now broken with similar errors. I guess they changed the structure of the API responses with the latest version. It would be great if our code could support both the old and new structures. If it's too much of a pain, then the only other option would be to drop support for the older versions. I prefer to keep the last 2 major versions around for testing, so that would be a last resort.

Thanks for the feedback. I've added a description to the PR.
The [API doc](https://acinq.github.io/eclair/#channels-2) doesn't have the sample responses for older versions. Had to run each versions while logging its responses to get that data. I've updated the PR to be compatible with the last 3 versions. Do take a look and let me know what you think. Thanks

jamaljsr commented:
Thanks for the updates. I will test it out this wknd.

jamaljsr reviewed (approved):
tACK LGTM 🚀

Thanks for the changes. This looks great! I tested all 3 versions and was able to open channels and make direct payments. 

There seems to be an issue with routing payments through Eclair v8.0+ nodes, but that is out of scope for this PR.
